src/lib/dataElements/javascriptTools.js

```diff
@@ -0,0 +1,103 @@
+/***************************************************************************************
+ * (c) 2018 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ ****************************************************************************************/
+
+'use strict';
+
+var isString = require('../helpers/stringAndNumberUtils').isString;
+
+/**
+ * The javascript tools data element.
+ * @param {Object} settings The data element settings object.
+ * @param {string} settings.operator The attribute that contains the desired operation.
+ * @returns {string}
+ */
+module.exports = function (settings) {
+  var value = settings.sourceValue;
+
+  switch (settings.operator) {
+    case 'simpleReplace':
+      value = String(value);
+      var method = settings.replaceAll === true ? 'replaceAll' : 'replace';
+      return value[method](settings.searchValue, settings.replacementValue);
+
+    case 'regexReplace':
+      value = String(value);
+      var re = new RegExp(
+        settings.regexInput,
+        (settings.caseInsensitive === true ? 'i' : '') +
+          (settings.replaceAll === true ? 'g' : '')
+      );
+
+      return value.replace(re, settings.replacementValue);
+
+    case 'substring':
+      value = String(value);
+      return value.substring(
+        settings.start,
+        settings.end != null ? settings.end : undefined
+      );
+
+    case 'regexMatch':
+      value = String(value);
+
+      var re = new RegExp(
+        settings.regexInput,
+        settings.caseInsensitive ? 'i' : ''
+      );
+
+      return value.match(re)[0];
+
+    case 'length':
+      if (!Array.isArray(value) && !isString(value)) {
+        return 0;
+      }
+
+      return value.length;
+
+    case 'slice':
+      if (!Array.isArray(value) && !isString(value)) {
+        value = String(value);
+      }
+
+      return value.slice(
+        settings.start,
+        settings.end != null ? settings.end : undefined
+      );
+
+    case 'indexOf':
+    case 'lastIndexOf':
+      value = String(value);
+      return value[settings.operator](settings.searchValue);
+
+    case 'join':
+      if (!Array.isArray(value)) {
+        return value;
+      }
+      return value.join(settings.delimiter);
+
+    case 'split':
```

Comment:
this will split `undefined`, which will be unexpected. Maybe use `isString` more liberally throughout here. Test for null and undefined in the tests?

Reply:
I will convert to string only the numbers. If the resulting value is not string, I will return the original value.